### PR TITLE
Enable failpoint in integration test

### PR DESCRIPTION
**Pull request by chaochn47**
Enable failpoint in integration test. Part of https://github.com/etcd-io/etcd/pull/15708 to address the original issue. 

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by chaochn47** [comment]
Could you please take a look? @serathius @ahrtr @jmhbnz @fuweid

File: Makefile [hunk 1]
```diff
@@ -25,15 +25,15 @@ $(TEMP_TEST_ANALYZER_DIR)/*:
 GO_TEST_FLAGS?=
 
 .PHONY: test
-test:
+test: gofail-enable
```
Comment (serathius) [inline Makefile:28]: Please don't add `gofail-enable` as a side effect of commands.
Just have integration test workflow enable it.

Comment (chaochn47) [reply Makefile:28]: It includes `integration` test workflow in the `make test` which is `PASSES="unit integration release e2e" ./scripts/test.sh $(GO_TEST_FLAGS)` :\ 

I tried make test but it failed with gofail is not enabled error message. How about I add to functional_test_pass? 

https://github.com/etcd-io/etcd/blob/ac61b36e069dd3daeca5d77257931a207b8d4545/scripts/test.sh#L122-L126

Comment (chaochn47) [reply Makefile]: Second thought, the specific test is skipped now. We can revisit this once it is enabled later.

Comment (serathius) [reply Makefile]: Like I said, do it in github workflow.

File: Makefile [hunk 2]
```diff
@@ -33,7 +33,7 @@ test-unit:
 	PASSES="unit" ./scripts/test.sh $(GO_TEST_FLAGS)
 
 .PHONY: test-integration
-test-integration:
+test-integration: gofail-enable
```
Comment (serathius) [inline Makefile:36]:
```suggestion
test-integration:
```

File: .github/workflows/tests-arm64.yaml [hunk 3]
```diff
@@ -35,6 +35,7 @@ jobs:
         run: |
           set -euo pipefail
           go clean -testcache
+          make gofail-enable
```
Comment (serathius) [inline .github/workflows/tests-arm64.yaml:38]: Note, not sure if we want to enable gofail in unit tests.

File: tests/integration/clientv3/cluster_test.go [hunk 4]
```diff
@@ -293,8 +293,11 @@ func TestMemberPromote(t *testing.T) {
 }
 
 // TestMemberPromoteMemberNotLearner ensures that promoting a voting member fails.
+// To test it locally, try run
+// make gofail-enable && pushd tests/integration/clientv3 && go test -v -run TestMemberPromoteMemberNotLearner && popd
```
Comment (serathius) [inline tests/integration/clientv3/cluster_test.go:297]: Not needed instructions.

File: tests/framework/integration/testing.go [hunk 5]
```diff
@@ -84,6 +98,13 @@ func BeforeTest(t testutil.TB, opts ...TestOption) {
 		testutil.RegisterLeakDetection(t)
 	}
 
+	if options.failpoint != nil && len(options.failpoint.name) != 0 {
+		require.NoError(t, gofail.Enable(options.failpoint.name, options.failpoint.payload))
```
Comment (serathius) [inline tests/framework/integration/testing.go:102]: What happens if gofail is not enabled? Could we skip the test here with a message that user should call `make gofail-enable`

File: .github/workflows/tests-arm64.yaml (hunk 3 above)
Comment (chaochn47) [reply .github/workflows/tests-arm64.yaml:38]: Thanks for reminder! I overlook it. Updating..

File: tests/integration/clientv3/cluster_test.go (hunk 4 above)
Comment (chaochn47) [reply tests/integration/clientv3/cluster_test.go]: ack

File: tests/integration/clientv3/cluster_test.go [hunk 6]
```diff
@@ -294,7 +294,8 @@ func TestMemberPromote(t *testing.T) {
 
 // TestMemberPromoteMemberNotLearner ensures that promoting a voting member fails.
 func TestMemberPromoteMemberNotLearner(t *testing.T) {
-	integration2.BeforeTest(t)
+	t.Skip("enable this test after PR 15708 is merged")
+	integration2.BeforeTest(t, integration2.WithFailpoint("raftBeforeAdvance", `sleep(100)`))
```
[the remark this answers is not in the document]
Comment (chaochn47) [reply tests/integration/clientv3/cluster_test.go:298]: sgtm. I will add a comment instead.

File: tests/framework/integration/testing.go [hunk 7]
```diff
@@ -84,6 +99,21 @@ func BeforeTest(t testutil.TB, opts ...TestOption) {
 		testutil.RegisterLeakDetection(t)
 	}
 
+	if options.failpoint != nil && len(options.failpoint.name) != 0 {
+		err := gofail.Enable(options.failpoint.name, options.failpoint.payload)
+		switch {
+		case err == gofail.ErrNoExist:
+			t.Skip(fmt.Sprintf("skip test %s due to error (%v); please run 'make gofail-enable' before running the test", t.Name(), err))
```
Comment (serathius) [inline tests/framework/integration/testing.go:106]:
```suggestion
			t.Skipf("skip test %s due to error (%v); please run 'make gofail-enable' before running the test", t.Name(), err)
```

File: tests/framework/integration/testing.go [hunk 8]
```diff
@@ -84,6 +99,21 @@ func BeforeTest(t testutil.TB, opts ...TestOption) {
 		testutil.RegisterLeakDetection(t)
 	}
 
+	if options.failpoint != nil && len(options.failpoint.name) != 0 {
+		err := gofail.Enable(options.failpoint.name, options.failpoint.payload)
+		switch {
+		case err == gofail.ErrNoExist:
```
Comment (serathius) [inline tests/framework/integration/testing.go:105]:
```suggestion
		case errors.Is(err, gofail.ErrNoExist):
```

Comment (serathius) [inline tests/framework/integration/testing.go:105]: Is there any other way to distinguish whether gofail is enabled other than checking if enabling failpoint succeeds?
`gofail.ErrNoExist` could be also returned if we made typo in failpoint name.

Comment (chaochn47) [reply tests/framework/integration/testing.go]: Yeah, there is a way. Once gofail is enabled, the failpoints map in runtime will be registered and `gofail/runtime.List()` will return an non-empty list. Updated in the next rev.

File: tests/framework/integration/testing.go (hunk 7 above)
Comment (chaochn47) [reply tests/framework/integration/testing.go]: There is no `t.Skipf`. I think it's overkill to add t.Name() given each test name will show up in the test output/report regardless. Updated in the latest rev.

**Review by ahrtr — APPROVED** [review]
LGTM

Thanks @chaochn47

File: tests/framework/integration/testing.go (hunk 8 above)
Comment (chaochn47) [reply tests/framework/integration/testing.go]: Yeah, that's the best practice of golang error comparing.. Thanks for pointing it out.

**Review by serathius — APPROVED** [review]

**Pull request merged**